addons/repo-proxy/common/src/main/java/org/commonjava/indy/repo/proxy/NPMMetadtaRewriteResponseDecorator.java

```diff
@@ -0,0 +1,220 @@
+        if ( isNPMMetaPath( path ) )
+        {
+            final String proxyToKeyPath = keyToPath( proxyToStoreKey );
+            trace( logger, "NPM rewriting replacement: from {} to {}", proxyToKeyPath, originalRepoPath );
+            return new NPMMetadataRewriteResponseWrapper( request, response, Collections.singletonMap( originalRepoPath,
+                                                                                                       proxyToKeyPath ) );
+        }
+        else
+        {
+            trace( logger, "NPM meta rewrite: {} is not a metadata path", path );
+        }
+
+        return response;
+    }
+
+    private boolean isNPMMetaPath( final String path )
+    {
+        if ( StringUtils.isBlank( path ) )
+        {
+            return false;
+        }
+        String checkingPath = path;
+        if ( path.startsWith( "/" ) )
+        {
+            checkingPath = path.substring( 1 );
+        }
+        // This is considering the single path for npm standard like "/jquery"
+        final boolean isSinglePath = checkingPath.split( "/" ).length < 2;
+        // This is considering the scoped path for npm standard like "/@type/jquery"
+        final boolean isScopedPath = checkingPath.startsWith( "@" ) && checkingPath.split( "/" ).length < 3;
+        // This is considering the package.json file itself
+        final boolean isPackageJson = checkingPath.trim().endsWith( "/" + METADATA_NAME );
+
+        trace( logger, "path: {}, isSinglePath: {}, isScopedPath: {}, isPackageJson: {}", path, isSinglePath,
+               isScopedPath, isPackageJson );
+        return isSinglePath || isScopedPath || isPackageJson;
+    }
+
+    private static class NPMMetadataRewriteResponseWrapper
+            extends HttpServletResponseWrapper
+    {
+        private HttpServletRequest request;
+
+        private NPMMetadataRewriteOutputStream out;
+
+        /**
+         * Constructs a response adaptor wrapping the given response.
+         * @throws IllegalArgumentException if the response is null
+         * @param response -
+         * @throws IOException -
+         */
+        public NPMMetadataRewriteResponseWrapper( final HttpServletRequest request, final HttpServletResponse response,
+                                                  final Map<String, String> reposReplacing )
+                throws IOException
+        {
+            super( response );
+            this.request = request;
+            this.out = new NPMMetadataRewriteOutputStream( response.getOutputStream(), reposReplacing );
+        }
+
+        @Override
+        public PrintWriter getWriter()
+                throws IOException
+        {
+            return new PrintWriter( this.getResponse().getWriter() );
+        }
+
+        @Override
+        public ServletOutputStream getOutputStream()
+        {
+            return this.out;
+        }
+
+    }
+
+    private static class NPMMetadataRewriteOutputStream
```

Comment:
I wonder if this would be useful to generalize for browsing output? Or, does the browsing use case just use the proxy redirection stuff we've already written?

Reply:
Yup agree. I'll pull this out as general one in the content browse rewrite feature pr.